models/files/xgbDART.R

```diff
@@ -0,0 +1,262 @@
+                                                nrounds = param$nrounds,
+                                                objective = "reg:linear",
+                                                booster= "dart",
+                                                ...)
+                    }
+                    out
+                    
+                  },
+                  predict = function(modelFit, newdata, submodels = NULL) {
+                    if(!inherits(newdata, "xgb.DMatrix")) {
+                      newdata <- as.matrix(newdata)
+                      newdata <- xgboost::xgb.DMatrix(data=newdata, missing = NA)
+                    }
+                    out <- predict(modelFit, newdata)
+                    if(modelFit$problemType == "Classification") {
+                      if(length(modelFit$obsLevels) == 2) {
+                        out <- ifelse(out >= .5,
+                                      modelFit$obsLevels[1],
+                                      modelFit$obsLevels[2])
+                      } else {
+                        out <- matrix(out, ncol = length(modelFit$obsLevels), byrow = TRUE)
+                        out <- modelFit$obsLevels[apply(out, 1, which.max)]
+                      }
+                    }
+                    
+                    if(!is.null(submodels)) {
+                      tmp <- vector(mode = "list", length = nrow(submodels) + 1)
+                      tmp[[1]] <- out
+                      for(j in seq(along = submodels$nrounds)) {
+                        tmp_pred <- predict(modelFit, newdata, ntreelimit = submodels$nrounds[j])
+                        if(modelFit$problemType == "Classification") {
+                          if(length(modelFit$obsLevels) == 2) {
+                            tmp_pred <- ifelse(tmp_pred >= .5,
+                                               modelFit$obsLevels[1],
+                                               modelFit$obsLevels[2])
+                          } else {
+                            tmp_pred <- matrix(tmp_pred, ncol = length(modelFit$obsLevels), byrow = TRUE)
+                            tmp_pred <- modelFit$obsLevels[apply(tmp_pred, 1, which.max)]
+                          }
+                        }
+                        tmp[[j+1]]  <- tmp_pred
+                      }
+                      out <- tmp
+                    }
+                    out
+                  },
+                  prob = function(modelFit, newdata, submodels = NULL) {
+                    if(!inherits(newdata, "xgb.DMatrix")) {
+                      newdata <- as.matrix(newdata)
+                      newdata <- xgboost::xgb.DMatrix(data=newdata, missing = NA)
+                    }
+                    
+                    if( !is.null(modelFit$param$objective) && modelFit$param$objective == 'binary:logitraw'){
+                      p <- predict(modelFit, newdata)
+                      out <- exp(p)/(1+exp(p))
```

Comment:
Using `out <- binomial()$linkinv(p)` would be better since it takes into account potential numerical issues